imap_processing/swe/l1a_swe.py

```diff
@@ -0,0 +1,183 @@
+    spin_three_indexes = [[3,  7, 11, 15, 19,  23], [21, 17, 13, 9, 5, 1]]
+    spin_four_indexes = [[4,  8, 12, 16, 20,  24], [22, 18, 14, 10, 6, 2]]
+
+    match spin_number:
+        case 0:
+            return spin_one_indexes
+        case 1:
+            return spin_two_indexes
+        case 2:
+            return spin_three_indexes
+        case 3:
+            return spin_four_indexes
+        case _:
+            print(f"Error! Invalid ESA_STEPS - {spin_number}")
+
+
+def swe_l1a():
+    """SWE L1A algorithm steps:
+        - Read data from SWE packet file
+        - Uncompress data
+        - Store metadata and data in attrs and DataArray of xarray respectively
+        - Save complete data to cdf file
+    Each L1A data will have this shape: 24 rows, 7 columns, and each cell in 24 x 7 table contains
+    30 element array. These dimension maps to this:
+        24 rows --> 24 energy steps
+        7 column --> 7 CEMs value
+        30 element --> 30 spin angles
+    """
+    packet_file = f"{packet_definition_directory}/../swe/tests/science_block_20221116_163611Z_idle.bin"
+    decom_data = decom_swe.decom_packets(packet_file)
+
+    total_packets = 0
+    one_full_cycle = np.zeros((24, 30, 7))
+    while total_packets < len(decom_data):
+        # If ESA_STEPS is 0, we are getting data for new cycle
+        spin_number = decom_data[total_packets].data["ESA_STEPS"].raw_value
+        if spin_number == 0:
+            # It should follow by 4 packets where each packet contains
+            # one spin's data.
+            one_full_cycle = np.zeros((24, 30, 7))
+
+        # TODO: get metadata of each data and stores. Find out how to combine four
+        # packet's metadata.
+
+        # read raw data
+        binary_data = decom_data[total_packets].data["SCIENCE_DATA"].raw_value
+        # read raw data as binary array using bitstring
+        bit_array = bitstring.ConstBitStream(bin=binary_data)
+        # chunk binary into 1260 units each with 8-bits
+        byte_data = bit_array.readlist(["uint:8"] * 1260)
+
+        # Uncompress counts. Uncompressed data is a list of 1260
+        # where 1260 = 15 seconds x 7 CEMs x 12 energy steps
+        uncompressed_data = [uncompress(i) for i in byte_data]
+        reshaped_data = np.array(uncompressed_data).reshape(15, 12, 7)
+        # TODO: Confirm how I am populating full data array with four spins data is correct.
+
+        # Reset column index for each spin.
+        column_index = 0
+        # Get indexes for each spin.
+        indexes = get_step_indexes(spin_number)
+
+        for every_second in range(15):
+            # go through first six steps of first half second and populate data in the right index
+            # of one_full_cycle array.
+            for every_step, index_number in enumerate(indexes[0]):
+                one_full_cycle[index_number-1, column_index] = reshaped_data[every_second, every_step]
```

Comment:
Wow, they didn't make this easy on you! Nice work decoding all this.

I _think_ you might be able to get rid of the outer `every_second` loop, because `column_index` is incrementing with `every_second`?

```suggestion
            for every_step, index_number in enumerate(indexes[0]):
                one_full_cycle[index_number-1, ::2] = reshaped_data[:, every_step]
                one_full_cycle[index_number-1, 1::2] = reshaped_data[:, 6 + every_step]
```

Reply:
I will try that. I was trying to think different ways and it was mind twisting. I was happy when this thing worked but was hoping for suggestion like this. I am going to ask them how they did in the past or what we have is good or not.

Reply:
If you're looking for more things to think about, I believe `take()` or `take_along_axis()` might be able to do the fancy numpy indexing for you to remove the Python loops altogether based on your index lookup tables.

https://numpy.org/doc/stable/reference/generated/numpy.take.html

I think it is very good to start with the basic for-loops here to make sure you have things working as you expect and then optimize one thing at a time, so I think you've got a good approach going right now.

Reply:
cool. I will take a look at that.